### ISSUE TITLE
RFE: move Provides/Requires generation to a separate container

### ISSUE THREAD

**Issue by nim-nim** [rpm-software-management/rpm#668]
The current `Provides`/`Requires` generation operates on a staging directory tree simulating the target deployment. This is mostly fine, except it breaks absolute symlinks within the staging directory. And, in some cases, those symlinks (and their target) participates in  `Provides`/`Requires` logic. So it needs special, sometimes invasive, logic inside the generation tool.

There is probably ways to do better nowadays by making generators operate on a container overlayed oved the build root, either in `rpmbuild` or just when the build is orchestrated by `mock`.

Since overhauling this stage to make use of modern system capabilities is likely to be tricky, this is a long-term should-eventually-be-done RFE.

**Comment by ignatenkobrain** [comment]
Do you have specific example where and how this should be used?

**Comment by nim-nim** [comment]
I the current way Go sources are organised, a "provides" is a directory containing Go source files. And projects are renamed all the time so it is common to alias previous names by symlinks (or http redirects in github)

so you have things like
```
/usr/share/gocode/src/oldname → /usr/share/gocode/src/newname 
```
(in newname package, for backwards compat)

At the stage generators operate on the symlink is broken
```
/prefix/usr/share/gocode/src/oldname → /usr/share/gocode/src/newname (which is in /prefix/usr/share/gocode/src/newname)
```

And the generator needs to walk the symlink to check it contains go code (recursively)

**Comment by ignatenkobrain** [comment]
This is wrong way to create symlink, you should always create relative ones. In that case it will work both during build and during runtime.

**Comment by pmatilai** [comment]
Yes, supporting bad packaging practises is not a reason to get into the container business. However people might not realize the implications of absolute symlinks in packages, it's something we could add a warning for.

**Comment by nim-nim** [comment]
It is a lot harder to create relative symlinks reliably in spec files, especially when part of the creating is given to manual packaging. (I simplified the example, you can have multiple levels of directories). The whole point of the FHS is not to have to worry about relative locations.

**Comment by ignatenkobrain** [comment]
@nim-nim 

from `ln(1)`:

```
       -r, --relative
              create symbolic links relative to link location
```

**Comment by pmatilai** [comment]
Another thought on the subject is that we could have a brp-script that turns absolute links into relative ones.

**Comment by nim-nim** [comment]
I need to re-check why `-r` was not used, whether there is a good reason or if it was just lost in the other breakage noise. Nevertheless that will only work (if it works) for the symlinks we generate ourselves, not those produced by upstream tools.

**Comment by ffesti** [comment]
One major issue here is that setting up containers/chroots/whatever typically requires root privileges. The build process is run as an unpriviledged user for a reason. Yes, rpmbuild could drop privileges but not even handnig them out is much more secure.

**Comment by pmatilai** [comment]
Renaming the RFE to something that might actually be worth implementing.

**Comment by nim-nim** [comment]
> I need to re-check why `-r` was not used, whether there is a good reason or if it was just lost in the other breakage noise. Nevertheless that will only work (if it works) for the symlinks we generate ourselves, not those produced by upstream tools.

Ok, I dig up why relative symlinks were given up on, and with several months of hindsight found how to workaround the problem. So we won't generate absolute symlinks at our packaging level. That still leaves the problem of upstream symlinks.

**Comment by Conan-Kudo** [comment]
`ln -r` didn't exist before sometime in 2014, I think? AFAIR, EL6 doesn't support it because of that...

**Comment by rwmjones** [comment]
Why are absolute symlinks a problem?  I don't understand why the warning added by @pmatilai is necessary.

**Comment by ffesti** [comment]
The problem with absolute smylinks is that they are pointing to system files outside the build root.

**Comment by ffesti** [comment]
This should most likely be implemented as a brp script.

**Comment by Conan-Kudo** [comment]
> Another thought on the subject is that we could have a brp-script that turns absolute links into relative ones.

This might be possible with the [`symlinks(1)`](https://www.mankier.com/1/symlinks) tool?

**Comment by Conan-Kudo** [comment]
Also, another example of this is [`dh_link`](https://salsa.debian.org/debian/debhelper/-/blob/main/dh_link).
